crates/weaver_semconv/src/deprecated.rs

```diff
@@ -0,0 +1,341 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! An enum to represent the different ways to deprecate an attribute, a metric, ...
+//!
+//! Two formats are supported:
+//! - A string with the deprecation message (old format)
+//! - A map with the action (renamed or removed) and optionally a note. When the
+//!   action is renamed, the map must also contain the field renamed_to.
+
+use regex::Regex;
+use schemars::JsonSchema;
+use serde::de::{MapAccess, Visitor};
+use serde::{de, Deserialize, Deserializer, Serialize};
+use std::fmt;
+use std::fmt::{Display, Formatter};
+
+/// The different ways to deprecate an attribute, a metric, ...
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "action")]
+pub enum Deprecated {
+    /// The object containing the deprecated field has been renamed to an
+    /// existing object or to a new object.
+    Renamed {
+        /// The new name of the field.
+        new_name: String,
+        /// An optional note to explain why the field has been renamed.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        note: Option<String>,
+    },
+    /// The object containing the deprecated field has been deprecated
+    /// either because it no longer exists, has been split into multiple fields,
+    /// has been renamed in various ways across different contexts, or for any other reason.
+    Deprecated {
+        /// A note to explain why the field has been deprecated.
+        note: String,
+    },
+}
+
+/// Custom deserialization function to handle both old and new formats.
+/// The old format is a string with the deprecation message.
+/// The new format is a map with the action (renamed or removed) and optionally a note. When the
+/// action is renamed, the map must also contain the field `rename_to`.
+pub fn deserialize_deprecated<'de, D>(deserializer: D) -> Result<Deprecated, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    // Define the visitor to handle both the old and new formats
+    struct DeprecatedVisitor;
+
+    impl<'de> Visitor<'de> for DeprecatedVisitor {
+        type Value = Deprecated;
+
+        fn expecting(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
+            formatter.write_str("a string or a map for deprecated field")
+        }
+
+        // Handle the old format (just a string)
+        fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            // Regex to match "Replaced by `some_field`"
+            // ToDo make this regex global
+            // Replaced by `([\w]+(?:\.[\w]+)*)`?
+            let renamed_regex =
```

Comment:
Do we want to keep this or just go update semconv to use structured data?

I'd be concerned we can't remove this once added.

Reply:
Currently, Weaver supports semconv registries from version 1.22 and above. For the `registry diff` command, I used a best-effort approach to support the `deprecated` field, but we could consider not doing so and instead fallback to `Deprecated::Deprecated` when the deserializer detects the old format. This would mean that meaningful diffs won’t be possible until we have two consecutive versions of the registry using the new format for `deprecated`.

I’m not opposed to removing this part of the code if you’re not comfortable with it. Please confirm.

Reply:
Barring comments from other @open-telemetry/specs-semconv-maintainers , I'm more comfortable keeping the weaver behavior here simple, and requiring explicit adoption of the new deprecation yaml for diffs to start working.

However if the ergonomics without this is quite bad, we can keep it, perhaps with some kind of flag we can disable it when in `--future` mode

Reply:
I removed the complex inference logic. To fully benefit from the `registry diff` command, we will need to migrate the registry to use the new `deprecated` format. That does seem more reasonable.